All_home_works/hw2/task1.py

```diff
@@ -0,0 +1,89 @@
+"""
+Given a file containing text. Complete using only default collections:
+    1) Find 10 longest words consisting from largest amount of unique symbols
+    2) Find rarest symbol for document
+    3) Count every punctuation char
+    4) Count every non ascii char
+    5) Find most common non ascii char for document
+"""
+import string
+from typing import List
+
+PUNCTUATION = set(string.punctuation)
+CROP_PUNCTUATION = str.maketrans({char: "" for char in PUNCTUATION})
+
+
+class Counter:
+    def __init__(self):
+        self.counter = {}
+
+    def add(self, key):
+        if key in self.counter:
+            self.counter[key] += 1
+        else:
+            self.counter[key] = 0
+
+    @property
+    def sort_counter(self):
+        sorted_counter = sorted(
+            self.counter.items(), key=lambda item: item[1], reverse=True
+        )
+        return list(sorted_counter)
+
+
+def _get_lines(file_path: str) -> str:
+    with open(file_path, encoding="raw_unicode_escape") as reading_file:
+        for line in reading_file:
+            line = line.rstrip("\n")
+            if line:
+                yield line
+
+
+def get_longest_diverse_words(file_path: str) -> List[str]:
+    queue = []
+    for line in _get_lines(file_path):
+        line = line.translate(CROP_PUNCTUATION)
+        words = [(word, len(set(word))) for word in line.split() if word]
+        queue = _check_words(queue, words)
+
+    return [word[0] for word in sorted(queue, key=lambda item: item[1], reverse=True)]
+
+
+def _check_words(queue, words):
+    for word, len_word in words:
+        if len(queue) < 10:
+            queue.append((word, len_word))
+            continue
+
+        index_word_len = min(enumerate(queue), key=lambda item: item[1][1])
+        if len_word > index_word_len[1][1]:
+            queue[index_word_len[0]] = (word, len_word)
+    return queue
+
+
+def get_rarest_char(file_path: str) -> str:
+    counter = Counter()
+    for line in _get_lines(file_path):
+        [counter.add(char) for char in line if char.isalpha()]
+    return list(reversed(counter.sort_counter))[0][0]
```

Comment:
What is the complexity of this line?  Can we do better?

Reply:
I have already answered this question in the next homework. So it can be done for O(n)

"If the most common element is needed, we can do it by O(n). But if we want to sort all elements n*log(n) needed"

Reply:
> I have already answered this question in the next homework.

I am not sure about timing, but when you fix something, I expect that you check and fix other similar cases as well.